src/game.cpp

```diff
@@ -5742,6 +5758,13 @@ int game::assign_faction_id()
     return ret;
 }
 
+long game::assign_creature_id()
+{
+    long ret = next_creature_id;
```

Comment:
Just adding one bit isn't going to buy enough space to make it really safe, you're still only guaranteed 32 bits, also I just remembered not just successive games but shared worlds need to be supported, take the above numbers and increase the creature generation rate by say 10x, and increase the likelihood of a large number of successive games by one or two orders of magnitude.   A big number will (mosly) just work, but it needs to be really big.
Another issue is concurrency with shared worlds, if there are multiple concurrent players this generator mostly stops doing anything to strictly avoid collisions.

A relatively simple solution is to use a long long, and reserve some bits for versioning and namespacing.  Something like stuff a version number in the first 8 bits of the number, stuff the player id in the next 16 bits of the number (limiting us to 65K distinct player ids, but that seems ok, right?) and the monster id goes in the next 47 bits (the rest).  You can detect monster rollover either by using MAX_LONG_LONG or by checking if it's negative after rollover.  If you want to recover a bit feel free to use an unsigned and detect rollover manually.  The first 24 bits would effectively constitute the player id, because that should be versioned too.
The version number lets us change schemes without worrying about collisions in continued games, and inserting the player id lets us not worry about concurrency, because creatures generated in different player's games will have disjoint sets of ids.